### PR TITLE
Composable stable exit

### PR DESCRIPTION
**Pull request by lgahdl**

### CODE REVIEW

File: src/entities/exit/types.ts
Comment (johngrantuk) [inline src/entities/exit/types.ts:1]: Could you do a similar refactor as we did on [Join types](https://github.com/balancer/b-sdk/pull/106/commits/6a08b5fbbb36e98e57bcbf02c903feba6617430c).

File: test/composableStableExit.integration.test.ts [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+      chainId,
+      rpcUrl,
+      bptIn,
+      kind: ExitKind.PROPORTIONAL,
+      exitWithNativeAsset: true,
+    };
+
+    // Note - checking native balance
+    const { queryResult, maxBptIn, minAmountsOut } = await doTransaction({
+      ...txInput,
```
Comment (johngrantuk) [inline test/composableStableExit.integration.test.ts:221]: Similar to join lets reuse any similar helper functions across tests.

File: src/entities/exit/poolExit.ts [hunk 2]
```diff
@@ -18,6 +19,7 @@ export class PoolExit {
         const { customPoolExits } = config || {};
         this.poolExits = {
             Weighted: new WeightedExit(),
+            PHANTOM_STABLE: new ComposableStableExit(),
```
Comment (johngrantuk) [inline src/entities/exit/poolExit.ts:22]: Can you add a comment (same as Join) that explains the `PHANTOM_STABLE` naming as its a bit confusing and useful for reference.

File: test/composableStableExit.integration.test.ts (hunk 1 above)
Comment (lgahdl) [reply test/composableStableExit.integration.test.ts:221]: But the composable doTransaction has a different expectedAmounts than the weighted, maybe I could use the same function, but adapt for each pool type?

Comment (johngrantuk) [reply test/composableStableExit.integration.test.ts]: I think there's a lot of similar functionality that could be reused?

File: src/entities/exit/composable-stable/composableStableExit.ts [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+            tokenOutIndex: amounts.tokenOutIndex,
+            toInternalBalance: !!input.toInternalBalance,
+            bptIndex,
+        };
+    }
+
+    private getAmountsQuery(
+        tokens: Token[],
+        input: ExitInput,
+        bptIndex?: number,
```
Comment (johngrantuk) [inline src/entities/exit/composable-stable/composableStableExit.ts:82]: We can make this non optional.

File: test/lib/utils/exitHelper.ts [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+        recipient: testAddress,
+    });
+
+    return {
+        exitBuildOutput,
+        exitQueryResult,
+    };
+};
+
+function getTokens(poolStateInput: PoolStateInput): Address[] {
```
Comment (johngrantuk) [inline test/lib/utils/exitHelper.ts:54]: we can reuse/split out from joinHelper?

File: test/lib/utils/exitHelper.ts [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+    // if exactIn slippage cannot be applied to bptIn, else should use bptIn with slippage applied
+    const maxBptIn = isExactIn
+        ? exitQueryResult.bptIn.amount
+        : slippage.applyTo(exitQueryResult.bptIn.amount);
+
+    const expectedBuildOutput: Omit<ExitBuildOutput, 'call'> = {
+        minAmountsOut,
+        maxBptIn,
+        to: BALANCER_VAULT,
+        // Value should equal value of any wrapped asset if using native
```
Comment (johngrantuk) [inline test/lib/utils/exitHelper.ts:341]:
```suggestion
        // Value should always be 0 for exits
```

**Review by johngrantuk — APPROVED** [review]
Looking good! Can go ahead and merge once the last little recent suggestions are updated.

File: test/lib/utils/exitHelper.ts (hunk 4 above)
Comment (lgahdl) [reply test/lib/utils/exitHelper.ts]: Yeah, it's the same function, I made it reusable, thanks!

**Pull request merged**